lib/optimizely/optimizely_user_context.rb

```diff
@@ -25,13 +25,14 @@ class OptimizelyUserContext
     attr_reader :user_id, :user_attributes
 
     def initialize(optimizely_client, user_id, user_attributes)
+      @set_attr_mutex = Mutex.new
       @optimizely_client = optimizely_client
       @user_id = user_id
       @user_attributes = user_attributes.nil? ? {} : user_attributes.clone
     end
 
     def set_attribute(attribute_key, attribute_value)
-      @user_attributes[attribute_key] = attribute_value
+      @set_attr_mutex.synchronize { @user_attributes[attribute_key] = attribute_value }
```

Comment:
We need the same lock protection for reading attributes as well.
Also can we return a snapshot copy of attributes when reading (so concurrent setAttribute won't change the attributes for ongoing decision)?
C# has the same issue being addressed (https://github.com/optimizely/csharp-sdk/pull/253).

Reply:
if we need to make sure its not modified while a decision is in progress, we will need to wrap all the userContext api methods in synchronize blocks. This will result in some slowdown on heavy use. I am not sure how making a copy could resolve this.. userContext is passes a few levels down the stack before user attributes are fetched. What do you think

Reply:
We do not need protect entire decide methods. The only possible conflict is to read/update attributes. 
You can see the charp-sdk PR linked above as a good reference.

Reply:
i just fixed it